File: picasso-sample/src/main/java/com/example/picasso/SampleGridViewAdapter.java
package com.example.picasso;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static android.widget.ImageView.ScaleType.CENTER_CROP;

final class SampleGridViewAdapter extends BaseAdapter {
  private final Context context;
  private final List<String> urls = new ArrayList<>();

  SampleGridViewAdapter(Context context) {
    this.context = context;

    // Ensure we get a different ordering of images on each run.
    Collections.addAll(urls, Data.URLS);
//    Collections.shuffle(urls);

    // Triple up the list.
    ArrayList<String> copy = new ArrayList<>(urls);
    urls.addAll(copy);
    urls.addAll(copy);
  }

  @Override public View getView(int position, View convertView, ViewGroup parent) {
    SquaredImageView view = (SquaredImageView) convertView;
    if (view == null) {
      view = new SquaredImageView(context);
      view.setScaleType(CENTER_CROP);
    }

    // Get the image URL for the current position.
    String url = getItem(position);

    // Trigger the download of the URL asynchronously into the image view.




    Picasso.get() //
        .load(url) //
        .placeholder(R.drawable.placeholder) //
        .error(R.drawable.error) //
//        .resize(400,200)
        .fit() //
        .tag(context) //
//        .centerCrop()
        .into(view, new Callback() {
          @Override
          public void onSuccess() {

          }

          @Override
          public void onError(Exception e) {

          }
        });

    return view;
  }

  @Override public int getCount() {
    return urls.size();
  }

  @Override public String getItem(int position) {
    return urls.get(position);
  }

  @Override public long getItemId(int position) {
    return position;
  }
}

File: picasso-sample/src/main/java/com/example/picasso/SampleGridViewAdapter.kt
/*
 * Copyright (C) 2022 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.picasso

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView.ScaleType.CENTER_CROP

internal class SampleGridViewAdapter(private val context: Context) : BaseAdapter() {
  private val urls: List<String>

  init {
    // Ensure we get a different ordering of images on each run.
    val tmpList = Data.URLS.toMutableList()
    tmpList.shuffle()

    // Triple up the list.
    urls = listOf(tmpList, tmpList, tmpList).flatten()
  }

  override fun getView(
    position: Int,
    convertView: View?,
    parent: ViewGroup
  ): View {
    val view = convertView as? SquaredImageView ?: SquaredImageView(context).apply {
      scaleType = CENTER_CROP
    }

    // Get the image URL for the current position.
    val url = getItem(position)

    // Trigger the download of the URL asynchronously into the image view.
    PicassoInitializer.get()
      .load(url)
      .placeholder(R.drawable.placeholder)
      .error(R.drawable.error)
      .fit()
      .tag(context)
      .into(view)

    return view
  }

  override fun getCount(): Int = urls.size

  override fun getItem(position: Int): String = urls[position]

  override fun getItemId(position: Int): Long = position.toLong()
}
